File: src/main/java/gargoyle/calendar/util/resources/Resources.java
package gargoyle.calendar.util.resources;

import java.io.IOException;
import java.io.OutputStream;
import java.net.URL;
import java.net.URLConnection;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.text.MessageFormat;

public final class Resources {

    public static final String PROTOCOL_FILE = "file";

    private Resources() {
    }


    public static Resource getResource(Path path) {
        return new FileSystemResource(path);
    }

    public static Resource getResource(String location, Class<?> aClass) {
        return new ClasspathResource(location, aClass);
    }

    public static Resource getResource(String location, ClassLoader loader) {
        return new ClasspathResource(location);
    }

    public static Resource getResource(URL url) {
        return new UrlResource(url);
    }


    public static OutputStream getURLStream(URL url) throws IOException {
        URLConnection connection = url.openConnection();
        return connection.getOutputStream();
    }


    public static OutputStream getFileStream(URL url) throws IOException {
        String location = url.toExternalForm();
        return Files.newOutputStream(Paths.get(location.substring(PROTOCOL_FILE.length() + 2)), StandardOpenOption.CREATE, StandardOpenOption.WRITE);
    }

    public static URL getUrl(String location, ClassLoader loader) {
        URL url = loader.getResource(location);
        if (url == null) throw new RuntimeIOException(MessageFormat.format("no resource {0}", location));
        return url;
    }

    public static URL getUrl(String location, Class<?> aClass) {
        URL url = aClass.getResource(location);
        if (url == null) url = aClass.getResource('/' + location);
        if (url == null) throw new RuntimeIOException(MessageFormat.format("no resource {0}", location));
        return url;
    }
}

File: cal-core/src/main/kotlin/gargoyle/calendar/util/resources/Resources.kt
package gargoyle.calendar.util.resources

import java.io.OutputStream
import java.net.URL
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardOpenOption
import kotlin.reflect.KClass

object Resources {
    const val PROTOCOL_FILE: String = "file"

    fun getResource(path: Path): Resource =
        FileSystemResource(path)

    fun getResource(location: String, type: KClass<*>): Resource =
        ClasspathResource(location, type)

    fun getResource(location: String, loader: ClassLoader): Resource =
        ClasspathResource(location, loader)

    fun getResource(url: URL): Resource =
        UrlResource(url)

    fun getURLStream(url: URL): OutputStream =
        url.openConnection().getOutputStream()

    fun getFileStream(url: URL): OutputStream =
        Files.newOutputStream(
            Paths.get(url.toExternalForm().substringAfter("$PROTOCOL_FILE:/")),
            StandardOpenOption.CREATE, StandardOpenOption.WRITE
        )

    fun getUrl(location: String, loader: ClassLoader): URL =
        requireNotNull(loader.getResource(location)) { "no resource $location" }

    fun getUrl(location: String, type: KClass<*>): URL =
        requireNotNull(type.java.getResource(location) ?: type.java.getResource("/$location"))
        { "no resource $location" }
}
